File: js/widget.ts
import type { RenderProps } from "@anywidget/types"
import { Cosmograph, CosmographConfig } from '@cosmograph/cosmograph'
import { tableFromIPC } from 'apache-arrow'

import { subscribe, toCamelCase } from './helper'
import { configProperties } from './config-props'

import "./widget.css"

function render({ model, el }: RenderProps) {
	el.classList.add("cosmograph_widget")
	const main = document.createElement('div')
  main.classList.add('widget-container')
	el.appendChild(main)

	const cosmographContainer = document.createElement('div')
  cosmographContainer.classList.add('cosmograph-container')
  main.appendChild(cosmographContainer)

	model.on('msg:custom', (msg: { [key: string]: never }) => {
		if (msg.type === 'select_point_by_index') {
			cosmograph?.selectPoint(msg.index, true)
		}
		if (msg.type === 'select_points_by_indices') {
			cosmograph?.selectPoints(msg.indices)
		}
		if (msg.type === 'activate_rect_selection') {
			cosmograph?.activateRectSelection()
		}
		if (msg.type === 'deactivate_rect_selection') {
			cosmograph?.deactivateRectSelection()
		}
		if (msg.type === 'fit_view') {
			cosmograph?.fitView()
		}
		if (msg.type === 'fit_view_by_indices') {
			cosmograph?.fitViewByIndices(msg.indices, msg.duration, msg.padding)
		}
		if (msg.type === 'fit_view_by_coordinates') {
			cosmograph?.fitViewByCoordinates(msg.coordinates, msg.duration, msg.padding)
		}
		if(msg.type === 'focus_point') {
			cosmograph?.focusPoint(msg.index ?? undefined)
		}
		if (msg.type === 'start') {
			cosmograph?.start(msg.alpha ?? undefined)
		}
		if (msg.type === 'pause') {
			cosmograph?.pause()
		}
		if (msg.type === 'restart') {
			cosmograph?.restart()
		}
		if (msg.type === 'step') {
			cosmograph?.step()
		}
	});

	const cosmographConfig: CosmographConfig = {
		onClick: (index) => {
			model.set('clicked_point_index', index)
			model.save_changes()
		},
		onPointsFiltered: () => {
			model.set('selected_point_indices', cosmograph?.getSelectedPointIndices() ?? [])
			model.save_changes()
		}
	}

	const modelChangeHandlers: { [key: string]: () => void } = {
		'change:_ipc_points': () => {
			const ipc = model.get('_ipc_points')
			cosmographConfig.points = ipc ? tableFromIPC(ipc.buffer) : []
		},
		'change:_ipc_links': () => {
			const ipc = model.get('_ipc_links')
			cosmographConfig.links = ipc ? tableFromIPC(ipc.buffer) : []
		}
	}

	// Set config properties from model
	configProperties.forEach(prop => {
		modelChangeHandlers[`change:${prop}`] = () => {
			const value = model.get(prop)

			// "disable_simulation" -> "disableSimulation", "simulation_decay" -> "simulationDecay", etc.
			if (value !== null) cosmographConfig[toCamelCase(prop) as keyof CosmographConfig] = value
		}
	})

	const unsubscribes = Object
		.entries(modelChangeHandlers)
		.map(([name, onModelChange]) => subscribe(model, name, () => {
			onModelChange()
			cosmograph.setConfig(cosmographConfig)
		}))

	// Initializes the Cosmograph with the configured settings
  Object.values(modelChangeHandlers).forEach(callback => callback())
	const cosmograph = new Cosmograph(cosmographContainer, cosmographConfig)

 return () => {
		unsubscribes.forEach(unsubscribe => unsubscribe())
	};
}

export default { render }

File: src/cosmograph_widget/widget.py
"""Cosmograph widget objects"""

import importlib.metadata
import pathlib
import json


import pyarrow as pa
import anywidget
from traitlets import Bool, Float, List, Int, Unicode, Union, Bytes, Any, observe

try:
    __version__ = importlib.metadata.version("cosmograph_widget")
except importlib.metadata.PackageNotFoundError:
    __version__ = "unknown"

meta_path = pathlib.Path(__file__).parent / "static" / "meta.json"
with open(meta_path) as f:
    meta_data = json.load(f)

class Cosmograph(anywidget.AnyWidget):
    _esm = pathlib.Path(__file__).parent / "static" / meta_data["js"]
    _css = pathlib.Path(__file__).parent / "static" / meta_data["css"]

    disable_simulation = Bool(None, allow_none=True).tag(sync=True)
    simulation_decay = Float(None, allow_none=True).tag(sync=True)
    simulation_gravity = Float(None, allow_none=True).tag(sync=True)
    simulation_center = Float(None, allow_none=True).tag(sync=True)
    simulation_repulsion = Float(None, allow_none=True).tag(sync=True)
    simulation_repulsion_theta = Float(None, allow_none=True).tag(sync=True)
    simulation_repulsion_quadtree_levels = Float(None, allow_none=True).tag(sync=True)
    simulation_link_spring = Float(None, allow_none=True).tag(sync=True)
    simulation_link_distance = Float(None, allow_none=True).tag(sync=True)
    simulation_link_dist_random_variation_range = List(
        None, default_value=None, allow_none=True
    ).tag(sync=True)
    simulation_repulsion_from_mouse = Float(None, allow_none=True).tag(sync=True)
    simulation_friction = Float(None, allow_none=True).tag(sync=True)

    background_color = Union(
        [Unicode(None, allow_none=True), List(Float, allow_none=True)]
    ).tag(sync=True)
    space_size = Int(None, allow_none=True).tag(sync=True)
    default_point_color = Union(
        [Unicode(None, allow_none=True), List(Float, allow_none=True)]
    ).tag(sync=True)
    point_greyout_opacity = Float(None, allow_none=True).tag(sync=True)
    default_point_size = Float(None, allow_none=True).tag(sync=True)
    point_size_scale = Float(None, allow_none=True).tag(sync=True)
    hovered_point_cursor = Unicode(None, allow_none=True).tag(sync=True)
    render_hovered_point_ring = Bool(None, allow_none=True).tag(sync=True)
    hovered_point_ring_color = Union(
        [Unicode(None, allow_none=True), List(Float, allow_none=True)]
    ).tag(sync=True)
    focused_point_ring_color = Union(
        [Unicode(None, allow_none=True), List(Float, allow_none=True)]
    ).tag(sync=True)
    focused_point_index = Int(None, allow_none=True).tag(sync=True)
    render_links = Bool(None, allow_none=True).tag(sync=True)
    default_link_color = Union(
        [Unicode(None, allow_none=True), List(Float, allow_none=True)]
    ).tag(sync=True)
    link_greyout_opacity = Float(None, allow_none=True).tag(sync=True)
    default_link_width = Float(None, allow_none=True).tag(sync=True)
    link_width_scale = Float(None, allow_none=True).tag(sync=True)
    curved_links = Bool(None, allow_none=True).tag(sync=True)
    curved_link_segments = Int(None, allow_none=True).tag(sync=True)
    curved_link_weight = Float(None, allow_none=True).tag(sync=True)
    curved_link_control_point_distance = Float(None, allow_none=True).tag(sync=True)
    default_link_arrows = Bool(None, allow_none=True).tag(sync=True)
    link_arrows_size_scale = Float(None, allow_none=True).tag(sync=True)
    link_visibility_distance_range = List(
        Float, default_value=None, allow_none=True
    ).tag(sync=True)
    link_visibility_min_transparency = Float(None, allow_none=True).tag(sync=True)
    use_quadtree = Bool(None, allow_none=True).tag(sync=True)
    show_FPS_monitor = Bool(None, allow_none=True).tag(sync=True)
    pixel_ratio = Float(None, allow_none=True).tag(sync=True)
    scale_points_on_zoom = Bool(None, allow_none=True).tag(sync=True)
    initial_zoom_level = Float(None, allow_none=True).tag(sync=True)
    disable_zoom = Bool(None, allow_none=True).tag(sync=True)
    enable_drag = Bool(None, allow_none=True).tag(sync=True)
    fit_view_on_init = Bool(None, allow_none=True).tag(sync=True)
    fit_view_delay = Float(None, allow_none=True).tag(sync=True)
    fit_view_padding = Float(None, allow_none=True).tag(sync=True)
    fit_view_duration = Float(None, allow_none=True).tag(sync=True)
    fit_view_by_points_in_rect = List(
        List(Float), default_value=None, allow_none=True
    ).tag(sync=True)
    random_seed = Union(
        [Int(None, allow_none=True), Unicode(None, allow_none=True)]
    ).tag(sync=True)
    point_sampling_distance = Int(None, allow_none=True).tag(sync=True)

    _ipc_points = Bytes().tag(sync=True)
    points = Any()

    _ipc_links = Bytes().tag(sync=True)
    links = Any()

    point_id = Unicode(None, allow_none=True).tag(sync=True)
    point_index = Unicode(None, allow_none=True).tag(sync=True)
    point_color = Unicode(None, allow_none=True).tag(sync=True)
    point_size = Unicode(None, allow_none=True).tag(sync=True)
    point_label = Unicode(None, allow_none=True).tag(sync=True)
    point_label_weight = Unicode(None, allow_none=True).tag(sync=True)
    point_x = Unicode(None, allow_none=True).tag(sync=True)
    point_y = Unicode(None, allow_none=True).tag(sync=True)
    point_include_columns = List(Unicode, default_value=None, allow_none=True).tag(
        sync=True
    )

    link_source = Unicode(None, allow_none=True).tag(sync=True)
    link_source_index = Unicode(None, allow_none=True).tag(sync=True)
    link_target = Unicode(None, allow_none=True).tag(sync=True)
    link_target_index = Unicode(None, allow_none=True).tag(sync=True)
    link_color = Unicode(None, allow_none=True).tag(sync=True)
    link_width = Unicode(None, allow_none=True).tag(sync=True)
    link_arrow = Unicode(None, allow_none=True).tag(sync=True)
    link_strength = Unicode(None, allow_none=True).tag(sync=True)
    link_include_columns = List(Unicode, default_value=None, allow_none=True).tag(
        sync=True
    )

    show_dynamic_labels = Bool(None, allow_none=True).tag(sync=True)
    show_labels_for = List(Unicode, allow_none=True).tag(sync=True)
    show_top_labels = Bool(None, allow_none=True).tag(sync=True)
    show_top_labels_limit = Int(None, allow_none=True).tag(sync=True)
    show_top_labels_by = Unicode(None, allow_none=True).tag(sync=True)
    static_label_weight = Float(None, allow_none=True).tag(sync=True)
    dynamic_label_weight = Float(None, allow_none=True).tag(sync=True)
    label_padding = Float(None, allow_none=True).tag(sync=True)
    show_hovered_point_label = Bool(None, allow_none=True).tag(sync=True)

    clicked_node_index = Int(None, allow_none=True).tag(sync=True)
    selected_point_indices = List(Int, allow_none=True).tag(sync=True)

    # Convert a Pandas DataFrame into a binary format and then write it to an IPC (Inter-Process Communication) stream.
    # The `with` statement ensures that the IPC stream is properly closed after writing the data.
    def get_buffered_arrow_table(self, df):
        # TODO: Add support for input data with different formats (e.g. CSV, Appache Arrow, DuckDB, etc.)
        table = pa.Table.from_pandas(df)
        sink = pa.BufferOutputStream()
        with pa.ipc.new_stream(sink, table.schema) as writer:
            writer.write(table)
        buffer = sink.getvalue()
        return buffer.to_pybytes()

    @observe("points")
    def change(self, change):
        points = change.new
        points_int32 = points.select_dtypes(include=['int64']).astype('int32')
        points[points_int32.columns] = points_int32
        self._ipc_points = self.get_buffered_arrow_table(points)

    @observe("links")
    def changeLinks(self, change):
        links = change.new
        links_int32 = links.select_dtypes(include=['int64']).astype('int32')
        links[links_int32.columns] = links_int32
        self._ipc_links = self.get_buffered_arrow_table(links)

    def select_point_by_index(self, index):
        self.send({"type": "select_point_by_index", "index": index})

    def select_points_by_indices(self, indices):
        self.send({"type": "select_points_by_indices", "indices": indices})

    def activate_rect_selection(self):
        self.send({"type": "activate_rect_selection"})

    def deactivate_rect_selection(self):
        self.send({"type": "deactivate_rect_selection"})

    def fit_view(self):
        self.send({"type": "fit_view"})

    def fit_view_by_indices(self, indices, duration=None, padding=None):
        self.send(
            {
                "type": "fit_view_by_indices",
                "indices": indices,
                "duration": duration,
                "padding": padding,
            }
        )

    def fit_view_by_coordinates(self, coordinates, duration=None, padding=None):
        self.send(
            {
                "type": "fit_view_by_coordinates",
                "coordinates": coordinates,
                "duration": duration,
                "padding": padding,
            }
        )

    def focus_point(self, index=None):
        self.send({"type": "focus_point", "index": index})

    def start(self, alpha=None):
        self.send({"type": "start", "alpha": alpha})

    def pause(self):
        self.send({"type": "pause"})

    def restart(self):
        self.send({"type": "restart"})

    def step(self):
        self.send({"type": "step"})
